lib/gala/payment_token.rb

```diff
@@ -37,6 +42,46 @@ def decrypt(certificate_pem, private_key_pem)
 
     class << self
 
+      def validate_signature(signature, ephemeral_public_key, data, transaction_id, application_data)
+        # Ensure that the certificates contain the correct custom OIDs
+        intermediate_cert = nil
+        leaf_cert = nil
+        p7 = OpenSSL::PKCS7.new(Base64.decode64(signature))
+        p7.certificates.each {|c|
+          c.extensions.each { |e|
+            leaf_cert = c if e.oid == LEAF_CERTIFICATE_OID
+            intermediate_cert = c if e.oid == INTERMEDIATE_CERTIFICATE_OID
+          }
+        }
+        raise InvalidSignatureError, "Signature does not contain the correct custom OIDs." unless leaf_cert && intermediate_cert
+
+        # Ensure that the root CA is the Apple Root CA - G3
+        root_cert = certificate = OpenSSL::X509::Certificate.new(APPLE_ROOT_CERT)
+
+        # Ensure that there is a valid X.509 chain of trust from the signature to the root CA
+        raise InvalidSignatureError, "Unable to verify a valid chain of trust from signature to root certificate." unless chain_of_trust_verified?(p7, leaf_cert, intermediate_cert, root_cert)
+
+        #Ensure that the signature is a valid ECDSA signature
+        unless application_data
+          verification_string = Base64.decode64(ephemeral_public_key) + Base64.decode64(data) + [transaction_id].pack("H*")
+          # verification_string = verification_string + application_data.pack("H*") if application_data
+          store = OpenSSL::X509::Store.new
+          verified = p7.verify([], store, verification_string, OpenSSL::PKCS7::NOVERIFY )
+          raise InvalidSignatureError, "The given signature is not a valid ECDSA signature." unless verified
+        end
+      end
+
+      def chain_of_trust_verified?(p7, leaf_cert, intermediate_cert, root_cert)
```

Comment:
Since the signature is verified here: https://github.com/spreedly/gala/pull/3/files#diff-e51c9557dcfb73c4dc1267c59d4358d7R69
I removed that redundancy and used the store since that looks to be the standard approach.